src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
+          evaluated (after optimization).
+        target: The target being viewed. Currently supports Point, OrientedPoint, and Object.
+        occludingObjects: An optional list of objects which can occlude the target.
+    """
+    occludingObjects = list(occludingObjects)
+
+    if isinstance(target, (Region, Object)):
+        # Extract the target region from the object or region.
+        if isinstance(target, Region):
+            raise NotImplementedError()
+        elif isinstance(target, (Object)):
+            target_region = target.occupiedSpace
+
+            # If the object contains its center and we can see the center, the object
+            # is visible.
+            if target_region.containsPoint(target.position) and \
+                canSee(position, orientation, visibleDistance, viewAngles, rayDensity, \
+                       target.position, occludingObjects):
+                return True
+
+        # Check that the distance to the target is not greater than visibleDistance, and
+        # see if position is in the target.
+        if target.containsPoint(position):
+            return True
+
+        if target.distanceTo(position) > visibleDistance:
+            return False
+
+        # Orient the object so that it has the same relative position and orientation to the
+        # origin as it did to the viewer
+        target_vertices = target_region.mesh.vertices - np.array(position.coordinates)
+        if orientation is not None:
+            target_vertices = orientation.invertRotation().getRotation().apply(target_vertices)
+
+        ## Check if the object crosses the y axis ahead and/or behind the viewer
+
+        # Extract the two vectors that are part of each edge crossing the y axis.
+        vec_0s = target_vertices[target_region.mesh.edges[:,0],:]
+        vec_1s = target_vertices[target_region.mesh.edges[:,1],:]
+        y_cross_edges = (vec_0s[:,0]/vec_1s[:,0]) < 0
+        vec_0s = vec_0s[y_cross_edges]
+        vec_1s = vec_1s[y_cross_edges]
+
+        # Figure out for which t value the vectors cross the y axis
+        t = (-vec_0s[:,0])/(vec_1s[:,0]-vec_0s[:,0])
+
+        # Figure out what the y value is when the y axis is crossed
+        y_intercept_points = t*(vec_1s[:,1]-vec_0s[:,1]) + vec_0s[:,1]
+
+        # If the object crosses ahead and behind the object, or through 0,
+        # we will not optimize ray casting.
+        target_crosses_ahead = np.any(y_intercept_points >= 0)
+        target_crosses_behind = np.any(y_intercept_points <= 0)
+
+        ## Compute the horizontal/vertical angle ranges which bound the object
+        ## (from the origin facing forwards)
+        spherical_angles = np.zeros((len(target_vertices[:,0]), 2))
+
+        spherical_angles[:,0] = np.arctan2(target_vertices[:,1], target_vertices[:,0])
+        spherical_angles[:,1] = np.arcsin(target_vertices[:,2]/ \
+                                (np.linalg.norm(target_vertices, axis=1)))
```

Comment:
Where possible, please use implicit line continuation with parentheses rather than explicit continuations with backslashes (they're easier to read).